scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -153,51 +195,97 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+            workers -= worker.run(self, p, end_idx, maxes, mins, _median, _compact, workers);
+            _less = build(self, start_idx, p, maxes, mins, _median, _compact, buflist, workers);
+            if(!worker)
+                _greater = build(self, p, end_idx, maxes, mins, _median, _compact, buflist, workers);
         }
         else
         {
             std::vector<double> tmp(m);
             double *mids = &tmp[0];
 
+            if(workers > 1) {
+                for (i=0; i<m; ++i) mids[i] = mins[i];
+                mids[d] = split;
+                
+                workers -= worker.run(self, p, end_idx, mids, mins, _median, _compact, workers);
+            }
+
             for (i=0; i<m; ++i) mids[i] = maxes[i];
             mids[d] = split;
-            _less = build(self, start_idx, p, mids, mins, _median, _compact);
+            _less = build(self, start_idx, p, mids, mins, _median, _compact, buflist, workers);
 
-            for (i=0; i<m; ++i) mids[i] = mins[i];
-            mids[d] = split;
-            _greater = build(self, p, end_idx, maxes, mids, _median, _compact);
+            if(!worker) {
+                for (i=0; i<m; ++i) mids[i] = mins[i];
+                mids[d] = split;
+                _greater = build(self, p, end_idx, maxes, mids, _median, _compact, buflist, workers);
+            }
         }
 
-        /* recompute n because std::vector can
-         * reallocate its internal buffer
-         */
-        root = tree_buffer_root(self->tree_buffer);
-        n = root + node_index;
+        n = &buf[node_index];
         /* fill in entries */
+        if(worker) {
+            worker.future_.wait();
+            _greater = 0;
+            for(const auto& b : *buflist)
+                _greater += b.size();
+            buflist->splice(buflist->end(), worker.buflist_);
+        }
         n->_less = _less;
         n->_greater = _greater;
-        n->less = root + _less;
-        n->greater = root + _greater;
         n->split_dim = d;
         n->split = split;
 
         return node_index;
     }
 }
 
-
-
 int build_ckdtree(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
-              double *maxes, double *mins, int _median, int _compact)
+              double *maxes, double *mins, int _median, int _compact, int _workers)
 
 {
-    build(self, start_idx, end_idx, maxes, mins, _median, _compact);
+    std::list<std::deque<ckdtreenode>> buflist;
+    buflist.emplace_back();
+    // build tree with sections collected in buflist
+    build(self, start_idx, end_idx, maxes, mins, _median, _compact, &buflist, _workers);
+
+    {
+        // count the number of nodes we got
+        ckdtree_intp_t tree_size = 0;
+        for(const auto& b : buflist) {
+            tree_size += b.size();
+        }
+        self->tree_buffer->resize(tree_size);
+    }
+
+    // merge sections from buflist in self->tree_buffer and update pointers and indices
+    std::deque<std::future<void>> futures;
```

Comment:
I will rather change this to call `futures.reserve(_workers)`, so that in the following loop I do not need to track an index and can used `push_back()`.

Reply:
Addressed in https://github.com/scipy/scipy/commit/fbb4a994ae81441c09baa299d8f9b0c1ce54f4ef .